lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,254 @@
+  std::set<parser::Label> labels() { return labels_; }
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> void Post(const T &) {}
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+  // C1136
+  void Post(const parser::ReturnStmt &) {
+    messages_.Say(currentStatementSourcePosition_,
+        "RETURN not allowed in DO CONCURRENT"_err_en_US);
+  }
+  // C1137
+  void NoImageControl() {
+    messages_.Say(currentStatementSourcePosition_,
+        "image control statement not allowed in DO CONCURRENT"_err_en_US);
+  }
+  void Post(const parser::SyncAllStmt &) { NoImageControl(); }
+  void Post(const parser::SyncImagesStmt &) { NoImageControl(); }
+  void Post(const parser::SyncMemoryStmt &) { NoImageControl(); }
+  void Post(const parser::SyncTeamStmt &) { NoImageControl(); }
+  void Post(const parser::ChangeTeamConstruct &) { NoImageControl(); }
+  void Post(const parser::CriticalConstruct &) { NoImageControl(); }
+  void Post(const parser::EventPostStmt &) { NoImageControl(); }
+  void Post(const parser::EventWaitStmt &) { NoImageControl(); }
+  void Post(const parser::FormTeamStmt &) { NoImageControl(); }
+  void Post(const parser::LockStmt &) { NoImageControl(); }
+  void Post(const parser::UnlockStmt &) { NoImageControl(); }
+  void Post(const parser::StopStmt &) { NoImageControl(); }
+  void Post(const parser::EndProgramStmt &) { NoImageControl(); }
+
+  void Post(const parser::AllocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "ALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+  }
+  void Post(const parser::DeallocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "DEALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+    // C1140: deallocation of polymorphic objects
+    if (anyObjectIsPolymorphic()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "DEALLOCATE polymorphic object(s) not allowed"
+          " in DO CONCURRENT"_err_en_US);
+    }
+  }
+  template<typename T> void Post(const parser::Statement<T> &) {
+    if (EndTDeallocatesCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "implicit deallocation of coarray not allowed"
+          " in DO CONCURRENT"_err_en_US);
+    }
+  }
+  // C1141: cannot call ieee_get_flag, ieee_[gs]et_halting_mode
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+      // C1137: call move_alloc with coarray arguments
+      if (name->ToString() == "move_alloc"s) {
+        if (anyObjectIsCoarray()) {
+          messages_.Say(currentStatementSourcePosition_,
```

Comment:
Can we point to the corarray object(s)?  Perhaps have N messages so we don't have to worry about the (s)?